### PR TITLE
fix(pattern.ts): enable use of regex flags for pattern in ruleset files

### PR DESCRIPTION
**Pull request by arno-di-loreto**
#242

### Please check if the PR fulfills these requirements
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce?

Bug fix

### What is the current behavior?

See #242

### Does this PR introduce a breaking change?

None

### CODE REVIEW

File: src/functions/pattern.ts [hunk 1]
```diff
@@ -1,5 +1,25 @@
 import { IFunction, IFunctionResult, IRulePatternOptions } from '../types';
 
+function test(value: string, regex: any) {
+  let re;
+  if (typeof regex === 'string') {
+    // regex in a string like {"match": "/[a-b]+/im"} or {"match": "[a-b]+"} in a json ruleset
+    // the available flags are "gimsuy" as described here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp
+    const splitRegex = /^\/(.+)\/([gimsuy]*)$/.exec(regex);
```
Comment (P0lip) [inline src/functions/pattern.ts]: I wouldn't be less lenient about flags here.
```suggestion
    const splitRegex = /^\/(.+)\/([a-z]*)$/.exec(regex);
```
IMHO we don't have to be so defensive here. Furthermore, we won't need to adjust the code in the future when a new flag is introduced. (`s` is fairly new for instance, introduced in ES2018 IIRC)

Moreover, we don't seem to allow regular expressions without any flags specified, so I'd suggest tweaking the above regular expression further
```suggestion
    const splitRegex = /^\/(.+)\/([a-z]*)?$/.exec(regex);
```

File: src/functions/pattern.ts [hunk 2]
```diff
@@ -1,5 +1,25 @@
 import { IFunction, IFunctionResult, IRulePatternOptions } from '../types';
 
+function test(value: string, regex: any) {
```
Comment (P0lip) [inline src/functions/pattern.ts]:
```suggestion
function test(value: string, regex: RegExp | string) {
```

File: src/functions/pattern.ts [hunk 3]
```diff
@@ -1,5 +1,25 @@
...
+  let re;
+  if (typeof regex === 'string') {
+    // regex in a string like {"match": "/[a-b]+/im"} or {"match": "[a-b]+"} in a json ruleset
+    // the available flags are "gimsuy" as described here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp
+    const splitRegex = /^\/(.+)\/([gimsuy]*)$/.exec(regex);
+    if (splitRegex) {
+      // with slashes like /[a-b]+/ and possibly with flags like /[a-b]+/im
+      re = new RegExp(splitRegex[1], splitRegex[2]);
```
Comment (P0lip) [inline src/functions/pattern.ts]: Syntax error caused by `/foo/iii` is uncaught here. Shall we wrap it with try...catch?
```suggestion
      try {
        re = new RegExp(splitRegex[1], splitRegex[2]);
      } catch {
        return false;
      }
```

Comment (arno-di-loreto) [reply src/functions/pattern.ts]: By silently catching errors, users wouldn't be aware that their regex is invalid. I would prefer to keep a behavior similar to what happens when you provide an invalid schema to the schema function (the exception is thrown, that really helped me 😅 ).
The following output shows how it looks when the schema and pattern function throw an exception:
```
Adding OpenAPI 2.0 (Swagger) functions
Encountered error when running rule 'invalid-regex-modifiers' on node at path '$,definitions':
SyntaxError: Invalid flags supplied to RegExp constructor 'izorglub'
Encountered error when running rule 'test-schema-error' on node at path '$,definitions,ValidSchemaName':
Error: schema is invalid: data/type should be equal to one of the allowed values, data/type should be array, data/type should match some schema in anyOf

/Users/arnaud/Dev/spectral-testing/samples/pattern-swagger.yaml
 23:24  error  schema-name-is-valid-not-working      Definitionname must not end with dto (any case)
 23:24  error  schema-name-is-valid-ulgy-workaround  Definitionname must not end with dto (any case)

✖ 2 problems (2 errors, 0 warnings, 0 infos)
```
If you're OK with that I just need to update the test to modify the one with an invalid flag. Can I do that on my branch and then use the `yarn commit` command without doing another pull request?

Comment (P0lip) [reply src/functions/pattern.ts]: > By silently catching errors, users wouldn't be aware that their regex is invalid. I would prefer to keep a behavior similar to what happens when you provide an invalid schema to the schema function (the exception is thrown, that really helped me sweat_smile ).

Agreed.

> If you're OK with that I just need to update the test to modify the one with an invalid flag. Can I do that on my branch and then use the yarn commit command without doing another pull request?

I'm totally fine with that. You can do it on your branch, sure.

File: src/functions/__tests__/pattern.test.ts [hunk 4]
```diff
@@ -48,10 +48,14 @@ describe('pattern', () => {
...
     });
 
-    test('should return an array with a message when given value matches the given match string regex with slashes and modifier', () => {
-      expect(runPattern('aBc', { match: '/[abc]+/izorglug' })).toEqual([
-        { message: "must match the pattern '/[abc]+/izorglug'" },
-      ]);
+    test('should throw an exception when given string regex contains invalid flags', () => {
+      let exception = null;
+      try {
+        runPattern('aBc', { match: '/[abc]+/invalid' });
```
Comment (P0lip) [inline src/functions/__tests__/pattern.test.ts]:
```ts
expect(() => runPattern('aBc', { match: '/[abc]+/invalid' })).toThrow(
  "Invalid flags supplied to RegExp constructor 'invalid'",
);
```

Could be simplified a bit :slightly_smiling_face:

**Comment by P0lip** [comment]
I went ahead and made the change. Hope you don't mind.

**Pull request merged**